cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -233,6 +235,43 @@ struct DivideChecked {
   }
 };
 
+struct Negate {
+  template <typename T, typename Arg0>
+  // NOTE [EPM]: Discuss on 0 vs. -0.
+  static constexpr enable_if_floating_point<T> Call(KernelContext*, Arg0 arg) {
+    return -arg;
+  }
+
+  // NOTE [EPM]: How to handle unsigned integers?
+  //  * Promote to signed?
+  //  * State that unsigned numbers are not supported (i.e., undefined behavior)?
+  //  * Use C++ integral conversions (e.g., Negate(-128) = -128)?
```

Comment:
To be clear, this is overflow behavior which is a separate concern from handling unsigned integers.

To match the behavior of the other arithmetic kernels, negation of 8 bit -128 should wrap around to -128 as you describe.

Reply:
For reference, numpy preserves the dtype for unsigned integers:

```
In [14]: arr = np.array([0,  255], dtype="uint8")

In [15]: -arr
Out[15]: array([0, 1], dtype=uint8)

In [16]: np.negative(arr)
Out[16]: array([0, 1], dtype=uint8)
```

(not sure that's very useful, though)